File: services/pii-service/app/detect.py
"""
PII detection for Belgian bill documents.

Identifies personal information in extracted text and returns
matches with their positions for user review before redaction.
"""

import re
from .models import PiiMatch

# Belgian national number: XX.XX.XX-XXX.XX or XXXXXXXXX-XX
NATIONAL_NUMBER_PATTERNS = [
    re.compile(r"\b\d{2}\.\d{2}\.\d{2}[-–]\d{3}\.\d{2}\b"),
    re.compile(r"\b\d{6}[-–]\d{3}[-–]\d{2}\b"),
    # NN without separators (11 digits starting with valid birth date)
    re.compile(r"\b[0-9]{2}[01]\d[0-3]\d\d{3}\d{2}\b"),
]

# Belgian phone numbers
PHONE_PATTERNS = [
    re.compile(r"(?:\+32|0032)\s*[\d\s./-]{8,12}"),
    re.compile(r"\b0\d[\s./\-]?\d{2,3}[\s./\-]?\d{2}[\s./\-]?\d{2}\b"),
    re.compile(r"\b04\d{2}[\s./\-]?\d{2}[\s./\-]?\d{2}[\s./\-]?\d{2}\b"),
]

# Payment-critical patterns to exclude from PII matches
IBAN_PATTERN = re.compile(r"\b[A-Z]{2}\d{2}\s*\d{4}\s*\d{4}\s*\d{4}\s*\d{0,4}\b")
STRUCTURED_COMM_PATTERN = re.compile(r"\+{3}\d{3}/\d{4}/\d{5}\+{3}")

# Email addresses
EMAIL_PATTERN = re.compile(r"\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}\b")

# Belgian address patterns
ADDRESS_PATTERNS = [
    # FR: Rue/Avenue/Boulevard + name + number
    re.compile(
        r"(?:Rue|Avenue|Boulevard|Place|Allée|Chemin|Impasse|Chaussée)"
        r"\s+[A-ZÀ-Ü][a-zà-ü]+(?:\s+[a-zà-ü]+)*\s+\d{1,4}(?:\s*[/,]\s*\d{1,4})?\b",
        re.IGNORECASE,
    ),
    # NL: straat/laan/weg/plein + number
    re.compile(
        r"\b[A-ZÀ-Ü][a-zà-ü]+(?:straat|laan|weg|plein|dreef|steenweg|singel)"
        r"\s+\d{1,4}(?:\s*[/,]\s*\d{1,4})?\b",
        re.IGNORECASE,
    ),
    # Postal code + city: 1000 Bruxelles, 1040 ETTERBEEK
    re.compile(
        r"\b[1-9]\d{3}[ \t]+[A-ZÀ-Ü][A-Za-zÀ-ü]+(?:[ \t]+[A-ZÀ-Ü][A-Za-zÀ-ü]+)?\b"
    ),
]

# Name headers in Belgian bills
NAME_HEADER_PATTERNS = [
    # Dutch
    re.compile(
        r"(?:Naam|Klant|Bestemming|Begunstigde|Geadresseerde)\s*:\s*[A-ZÀ-Ü][^\n,]{2,40}",
        re.IGNORECASE,
    ),
    # French
    re.compile(
        r"(?:Nom|Client|Destinataire|Titulaire)\s*:\s*[A-ZÀ-Ü][^\n,]{2,40}",
        re.IGNORECASE,
    ),
    # Standalone all-caps name on its own line (e.g. "AMIR SIBAT" in addressee block)
    # Must be 2-4 words, each 2+ uppercase letters, on its own line
    re.compile(
        r"^[ \t]*([A-ZÀ-Ü]{2,}(?:[ \t]+[A-ZÀ-Ü]{2,}){1,3})[ \t]*$",
        re.MULTILINE,
    ),
]

# Words that look like all-caps names but are not (bill labels, headers)
NAME_EXCLUSIONS = {
    "AVANT LE", "TOTAL", "IBAN", "BIC", "TVA", "TVAC", "HTVA",
    "MONTANT", "EUR", "FACTURE", "UNE QUESTION", "PAYER",
}

REPLACEMENT_MAP: dict[str, str] = {
    "national_number": "[REDACTED_NN]",
    "phone": "[REDACTED_PHONE]",
    "email": "[REDACTED_EMAIL]",
    "address": "[REDACTED_ADDRESS]",
    "name_header": "[REDACTED_NAME]",
}


def _find_matches(
    text: str, patterns: list[re.Pattern[str]], pii_type: str
) -> list[PiiMatch]:
    matches: list[PiiMatch] = []
    for pattern in patterns:
        for m in pattern.finditer(text):
            matches.append(
                PiiMatch(
                    type=pii_type,
                    value=m.group(),
                    start=m.start(),
                    end=m.end(),
                    replacement=REPLACEMENT_MAP[pii_type],
                )
            )
    return matches


def _deduplicate_overlapping(matches: list[PiiMatch]) -> list[PiiMatch]:
    if len(matches) <= 1:
        return matches

    result: list[PiiMatch] = [matches[0]]
    for i in range(1, len(matches)):
        prev = result[-1]
        curr = matches[i]
        if curr.start < prev.end:
            # Overlapping — keep the longer match
            if (curr.end - curr.start) > (prev.end - prev.start):
                result[-1] = curr
        else:
            result.append(curr)
    return result


def _find_protected_ranges(text: str) -> list[tuple[int, int]]:
    """Find ranges of payment-critical fields that should never be flagged as PII."""
    ranges: list[tuple[int, int]] = []
    for m in IBAN_PATTERN.finditer(text):
        ranges.append((m.start(), m.end()))
    for m in STRUCTURED_COMM_PATTERN.finditer(text):
        ranges.append((m.start(), m.end()))
    return ranges


def _overlaps_protected(match: PiiMatch, protected: list[tuple[int, int]]) -> bool:
    for start, end in protected:
        if match.start < end and match.end > start:
            return True
    return False


def detect_pii(text: str) -> list[PiiMatch]:
    """Detect PII in text. Returns all matches sorted by position."""
    matches: list[PiiMatch] = []

    matches.extend(_find_matches(text, NATIONAL_NUMBER_PATTERNS, "national_number"))
    matches.extend(_find_matches(text, PHONE_PATTERNS, "phone"))
    matches.extend(_find_matches(text, [EMAIL_PATTERN], "email"))
    matches.extend(_find_matches(text, ADDRESS_PATTERNS, "address"))
    matches.extend(_find_matches(text, NAME_HEADER_PATTERNS, "name_header"))

    # Filter out all-caps "names" that are actually bill labels
    matches = [
        m for m in matches
        if not (m.type == "name_header" and m.value.strip() in NAME_EXCLUSIONS)
    ]

    # Filter out matches that overlap with IBANs or structured communications
    protected = _find_protected_ranges(text)
    matches = [m for m in matches if not _overlaps_protected(m, protected)]

    matches.sort(key=lambda m: m.start)
    return _deduplicate_overlapping(matches)


def apply_redactions(
    text: str,
    matches: list[PiiMatch],
    approved_indices: list[int] | None = None,
) -> str:
    """Apply redactions to text. Only redacts approved matches (or all if none specified)."""
    to_redact = (
        [m for i, m in enumerate(matches) if i in approved_indices]
        if approved_indices is not None
        else matches
    )

    # Apply from end to start to preserve positions
    result = text
    for m in reversed(to_redact):
        result = result[: m.start] + m.replacement + result[m.end :]
    return result

File: lib/pii/detect.ts
/**
 * PII detection for Belgian bill documents.
 *
 * Identifies personal information in extracted text and returns
 * matches with their positions for user review before redaction.
 */

export type PiiType =
  | 'national_number'  // Belgian NN: XX.XX.XX-XXX.XX
  | 'phone'            // Belgian phone: +32..., 04XX/XX.XX.XX
  | 'email'            // email@domain.com
  | 'address'          // Street + number patterns
  | 'name_header'      // "Naam/Nom:" followed by a name

export interface PiiMatch {
  type: PiiType
  value: string
  start: number
  end: number
  replacement: string
}

/** Belgian national number: XX.XX.XX-XXX.XX or XXXXXXXXX-XX */
const NATIONAL_NUMBER_PATTERNS = [
  /\b\d{2}\.\d{2}\.\d{2}[-–]\d{3}\.\d{2}\b/g,
  /\b\d{6}[-–]\d{3}[-–]\d{2}\b/g,
  // NN without separators (11 digits starting with valid birth date)
  /\b[0-9]{2}[01]\d[0-3]\d\d{3}\d{2}\b/g,
]

/** Belgian phone numbers */
const PHONE_PATTERNS = [
  /(?:\+32|0032)\s*[\d\s./-]{8,12}/g,
  /\b0\d[\s./\-]?\d{2,3}[\s./\-]?\d{2}[\s./\-]?\d{2}\b/g,
  /\b04\d{2}[\s./\-]?\d{2}[\s./\-]?\d{2}[\s./\-]?\d{2}\b/g,
]

/** Email addresses */
const EMAIL_PATTERN = /\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}\b/g

/**
 * Belgian address patterns — street names followed by numbers.
 * Matches common patterns like:
 *   "Rue de la Loi 16", "Wetstraat 16", "Avenue Louise 123/4"
 */
const ADDRESS_PATTERNS = [
  // FR: Rue/Avenue/Boulevard + name + number
  /(?:Rue|Avenue|Boulevard|Place|Allée|Chemin|Impasse|Chaussée)\s+[A-ZÀ-Ü][a-zà-ü]+(?:\s+[a-zà-ü]+)*\s+\d{1,4}(?:\s*[\/,]\s*\d{1,4})?\b/gi,
  // NL: straat/laan/weg/plein + number
  /\b[A-ZÀ-Ü][a-zà-ü]+(?:straat|laan|weg|plein|dreef|steenweg|singel)\s+\d{1,4}(?:\s*[\/,]\s*\d{1,4})?\b/gi,
  // Postal code + city: 1000 Bruxelles, 2000 Antwerpen
  /\b[1-9]\d{3}[ \t]+[A-ZÀ-Ü][a-zà-ü]+(?:[ \t]+[A-ZÀ-Ü][a-zà-ü]+)?\b/g,
]

/** Name headers in Belgian bills */
const NAME_HEADER_PATTERNS = [
  // Dutch: "Naam:", "Naam van de klant:", "Klant:"
  /(?:Naam|Klant|Bestemming|Begunstigde|Geadresseerde)\s*:\s*[A-ZÀ-Ü][^\n,]{2,40}/gi,
  // French: "Nom:", "Nom du client:", "Destinataire:"
  /(?:Nom|Client|Destinataire|Titulaire)\s*:\s*[A-ZÀ-Ü][^\n,]{2,40}/gi,
]

const REPLACEMENT_MAP: Record<PiiType, string> = {
  national_number: '[REDACTED_NN]',
  phone: '[REDACTED_PHONE]',
  email: '[REDACTED_EMAIL]',
  address: '[REDACTED_ADDRESS]',
  name_header: '[REDACTED_NAME]',
}

/**
 * Detect PII in text. Returns all matches sorted by position.
 * Does NOT redact — returns matches for user review.
 */
export function detectPii(text: string): PiiMatch[] {
  const matches: PiiMatch[] = []

  // National numbers
  for (const pattern of NATIONAL_NUMBER_PATTERNS) {
    for (const match of text.matchAll(new RegExp(pattern))) {
      matches.push({
        type: 'national_number',
        value: match[0],
        start: match.index!,
        end: match.index! + match[0].length,
        replacement: REPLACEMENT_MAP.national_number,
      })
    }
  }

  // Phones
  for (const pattern of PHONE_PATTERNS) {
    for (const match of text.matchAll(new RegExp(pattern))) {
      matches.push({
        type: 'phone',
        value: match[0],
        start: match.index!,
        end: match.index! + match[0].length,
        replacement: REPLACEMENT_MAP.phone,
      })
    }
  }

  // Emails
  for (const match of text.matchAll(new RegExp(EMAIL_PATTERN))) {
    matches.push({
      type: 'email',
      value: match[0],
      start: match.index!,
      end: match.index! + match[0].length,
      replacement: REPLACEMENT_MAP.email,
    })
  }

  // Addresses
  for (const pattern of ADDRESS_PATTERNS) {
    for (const match of text.matchAll(new RegExp(pattern))) {
      matches.push({
        type: 'address',
        value: match[0],
        start: match.index!,
        end: match.index! + match[0].length,
        replacement: REPLACEMENT_MAP.address,
      })
    }
  }

  // Name headers
  for (const pattern of NAME_HEADER_PATTERNS) {
    for (const match of text.matchAll(new RegExp(pattern))) {
      matches.push({
        type: 'name_header',
        value: match[0],
        start: match.index!,
        end: match.index! + match[0].length,
        replacement: REPLACEMENT_MAP.name_header,
      })
    }
  }

  // Filter out matches that overlap with IBANs or structured communications
  const protectedRanges = findProtectedRanges(text)
  const filtered = matches.filter(m => !protectedRanges.some(([s, e]) => m.start < e && m.end > s))

  // Sort by position and deduplicate overlapping matches
  filtered.sort((a, b) => a.start - b.start)
  return deduplicateOverlapping(filtered)
}

const IBAN_RE = /\b[A-Z]{2}\d{2}\s*\d{4}\s*\d{4}\s*\d{4}\s*\d{0,4}\b/g
const STRUCTURED_COMM_RE = /\+{3}\d{3}\/\d{4}\/\d{5}\+{3}/g

function findProtectedRanges(text: string): [number, number][] {
  const ranges: [number, number][] = []
  for (const m of text.matchAll(new RegExp(IBAN_RE))) {
    ranges.push([m.index!, m.index! + m[0].length])
  }
  for (const m of text.matchAll(new RegExp(STRUCTURED_COMM_RE))) {
    ranges.push([m.index!, m.index! + m[0].length])
  }
  return ranges
}

/**
 * Remove overlapping matches, keeping the longer one.
 */
function deduplicateOverlapping(matches: PiiMatch[]): PiiMatch[] {
  if (matches.length <= 1) return matches

  const result: PiiMatch[] = [matches[0]]
  for (let i = 1; i < matches.length; i++) {
    const prev = result[result.length - 1]
    const curr = matches[i]
    if (curr.start < prev.end) {
      // Overlapping — keep the longer match
      if (curr.end - curr.start > prev.end - prev.start) {
        result[result.length - 1] = curr
      }
    } else {
      result.push(curr)
    }
  }
  return result
}

/**
 * Apply redactions to text based on PII matches.
 * Only redacts matches the user has approved (by index).
 */
export function applyRedactions(text: string, matches: PiiMatch[], approvedIndices?: number[]): string {
  // If no approved indices specified, redact all
  const toRedact = approvedIndices
    ? matches.filter((_, i) => approvedIndices.includes(i))
    : matches

  // Apply from end to start to preserve positions
  let result = text
  for (let i = toRedact.length - 1; i >= 0; i--) {
    const m = toRedact[i]
    result = result.slice(0, m.start) + m.replacement + result.slice(m.end)
  }
  return result
}
